text/0000-generalized-type-ascription.md

```diff
@@ -0,0 +1,1563 @@
+
+[RFC_803_motivation]: https://github.com/rust-lang/rfcs/blob/master/text/0803-type-ascription.md#motivation
+
+[pointfree persuasion]: https://en.wikipedia.org/wiki/Tacit_programming
+
+[TwoHardThings]: https://martinfowler.com/bliki/TwoHardThings.html
+
+Type ascription is useful. A motivation for the feature is noted in the merged,
+but thus far not stabilized, [RFC 803][RFC_803_motivation] which introduces
+type ascription in expression contexts as `expr : T`. We reinforce that RFC
+with more motivation:
+
+1. Introducing a temporary `let` binding of form `let ident: type = expr;`,
+   as a substitute for type-ascription, forces programmers to invent artificial
+   variable names. Naming is hard (particularly for those who are of the
+   [pointfree persuasion]). As [the saying goes][TwoHardThings]:
+   > “There are only two hard things in Computer Science: cache invalidation and naming things”.
+
+   By reducing the pressure on Rustaceans to name artificial units
+   we can let programmers focus on naming where it matters more (API boundaries).
+
+2. Type ascription helps retain writing flow.
+   When you are writing a complex chain of methods,
+   sometimes you realize that you need to add an annotation,
+   either to make things compile or for the purposes of documentation.
+   When you do that, it follows the flow of writing the method chain
+   to not have to split things into let bindings; instead,
+   you can simply add an annotation to the right of a method call in the chain
+   and then continue on with the next method call.
+
+   Similarly, type ascription also follows the reading flow well and does so
+   in a non-intrusive way.
+
+3. When you want to pass an expression such as `try { .. }` or
+   `async { .. }` to a function which expects a generic `R: Try` or `R: Future`,
+   type inference can fail. In this case,
+   it is more ergonomic to type-ascribe with `try : R { .. }` instead
+   of first introducing an artificial binding.
+
+4. With type ascription, you can annotate smaller bits and subsets of what you
+   previously needed to. This especially holds in pattern contexts.
+   This will be made clear later on in this RFC.
+
+5. Turbofish is not always possible! Consider for example:
+
+   ```rust
+   fn display_all(elems: impl Iterator<Item: Display>) { ... }
+   ```
+
+   As of today (2018-08-06), it is not possible to use turbofish at all
+   as you could have done had you instead written:
+
+   ```rust
+   fn display_all<I: Iterator<Item: Display>>(elems: I) { ... }
+   ```
+
+   While this may change in the future, it may also be the case that anonymous
+   `arg: impl Trait`s can't ever be turbofished. In such a case, type ascription
+   is our saving grace; it works independently of how `display_all` is defined
+   and let's us easily constrain the type of `elems` in a syntactically
+   light-weight way in any case.
+
+6. Type ascription is helpful when doing *type* driven development (TDD)
```

Comment:
nit: When I google "TDD" I get _test_ driven development, as I expected, so I would discourage this abbreviation.